doc/sphinx-guides/source/developers/s3-direct-upload-api.rst

```diff
@@ -102,7 +102,25 @@ If the client is unable to complete the multipart upload, it should call the abo
   
     curl -X DELETE "$SERVER_URL/api/datasets/mpload?..."
    
-  
+
+.. _direct-async-upload:
+
+Asynchronous Direct Upload
```

Comment:
I'm confused by this being a section and having a curl example. This isn't really a separate step or kind of upload (as far as S3 is concerned) is it? And the curl command above works? (Or does it fail for some S3 services?)

If I understand, you're seeing this from python - could it be that w/o setting the size in python, when using the async library, you switch to chunked encoding which requires HTTP 1.1 which may not be implemented somewhere (what S3 service are you testing with?)? (I'm not sure if this is what you're seeing - trying to Google quickly. In Java, the DVUploader uses Apache's InputStreamEntity(stream, length) and I've seen an issue where someone tried -1 as a length and the answer reported that -1 switches to chunked mode which requires HTTP 1.1 and they saw a not implemented error)

If it's something like this, I'd suggest adding this as a Note: on the file upload step (which starts at line 74 but doesn't have a header right now for some reason) and either noting that for some servers a size header is required (if the original curl isn't working) or noting that, when writing code, adding the size is required/may be required depending on the S3 service (if it is about HTTP1.1 support) or just noting that if you see a 501 error when developing code to implement this, assuring that a size header is sent is one possible way to resolve it (if we really don't know the underlying cause).  I'd suggest only including a curl example if there are cases where the original curl fails.

Reply:
Thank you for the quick review, @qqmyers. 

I’ve encountered this issue not only in Python (httpx, aiohttp/aiofiles) but also in Rust (reqwest), and I suspect this might vary between libraries. Hence, the idea was to add it as a note or section in the docs to provide troubleshooting for those who may experience the same issue.

> Which S3 service are you testing with?

I’ve run into this issue while testing `python-dvuploader` and `dataverse-rs` on demo.dataverse.org. It’s also been reported in another installation as detailed in https://github.com/gdcc/python-dvuploader/issues/7. However, I’m not certain about the specific S3 service involved. Interestingly, using the `localstack` service on a local instance didn’t trigger the issue, which can create a misleading sense of security since tests are not failing.

> If it’s something like this, I’d suggest adding this as a Note: on the file upload step […]

I agree, that sounds like a more appropriate solution. I initially added it as a section because I thought developers integrating the Direct Upload API might be the primary audience, and I wanted to ensure it stands out. I would condense the current section into a note without the `curl` example.

Reply:
Something along these lines? 

```
.. note::
    If you encounter an ``HTTP 501 Not Implemented`` error, 
    ensure the `Content-Length` header is correctly set to 
    the file or chunk size. This issue may arise when streaming 
    files or chunks asynchronously to S3 via ``PUT`` requests, 
    particularly if the library or tool you're using doesn't
    set the ``Content-Length`` header automatically.
```

Ignore the weird newlines. I just tried to fit it into this comment 😁

Reply:
Sounds great - could you also add the ~missing heading before line 74 as well - just "Upload Files to S3" or similar.  (Weird that I said three steps and step 2 doesn't have a header...)

Reply:
Sounds good! For the missing section header, should I add `.. _direct-upload-to-s3:` before the "Upload Files to S3" section to ensure it appears in the TOC?

Reply:
@qqmyers - I have updated the PR

Reply:
Looks good to me. Thanks.